codes/extractive/netsum/netsum.py

```diff
@@ -0,0 +1,219 @@
+
+@author: Rose
+"""
+from __future__ import division
+import nltk
+from nltk.tokenize import sent_tokenize, word_tokenize, RegexpTokenizer
+import pandas as pd
+import numpy as np
+from collections import Counter
+import h5py
+import os
+import keras
+from sklearn.model_selection import train_test_split
+from keras.models import Sequential, Model
+from keras.layers import Dense, Dropout, Activation, Flatten, Conv2D, MaxPooling2D, Input
+from keras.callbacks import ModelCheckpoint, EarlyStopping
+
+text_path = "/root/nlp_project/cnn/text_2000/"
+summary_path = "/root/nlp_project/cnn/summary_2000/"
+file_list = "/root/nlp_project/cnn/datalist2.csv"
+
+# min. input size: 30 (sentences)
+MAX_INPUT_SEQ_LENGTH=30
+BATCH_SIZE = 20
+EPOCHS = 20
+
+# load the data
+def data_prep(text_path, summary_path, file_info, save_transformed_flag, transformed_path=None):
+    label = []
+    feature= []
+    data_df = pd.read_csv(file_list,sep=',',header='infer')
+    word_tokenizer = RegexpTokenizer(r'\w+')
+    
+    for i in range(len(data_df)):
+        label_list = []
+        feature_list = []
+        padding_needed = False
+        text_file = os.path.join(text_path,data_df["text_path"][i])
+        with open(text_file, 'r', encoding="utf8") as rf:
+            text = rf.read()
+            text_sents = sent_tokenize(text)
+            all_docs_words = word_tokenizer.tokenize(text)
+            text_word_counter = Counter(all_docs_words)
+            all_docs_bigram = list(nltk.bigrams(all_docs_words))
+            total_num_words = len(all_docs_words)
+    
+        summary_file = os.path.join(summary_path,data_df["summary_path"][i])
+        with open(summary_file, 'r', encoding="utf8") as rf:
+            text = rf.read()
+            summary_array = word_tokenizer.tokenize(text)
+        
+        # focing the same shape.
+        if len(text_sents) > MAX_INPUT_SEQ_LENGTH:
+            text_sents = text_sents[:MAX_INPUT_SEQ_LENGTH]
+        else:
+            padding_needed = True
+            
+        # Calculating rouge1 label, as described in the paper.
+        for j in range(len(text_sents)):
+            both_occur = 0
+            cum_prob_word = 0
+            cum_prob_bigram = 0
+            word_array = word_tokenizer.tokenize(text_sents[j])
+            all_sent_bigram = list(nltk.bigrams(word_array))
+            for w in word_array:
+                if w in summary_array:
+                    both_occur += 1
+                cum_prob_word += text_word_counter[w] / total_num_words
+            for bg in all_sent_bigram:
+                cum_prob_bigram += all_docs_bigram.count(bg) / len(all_docs_bigram)
+            # get the rouge1 score here and append to the vector.
+            label_list.append(both_occur/len(summary_array))
+            # get other X features here.
+            is_first_sentence = 0
+            if j == 0:
+                is_first_sentence = 1
+            sent_position = j / len(text_sents)
```

Comment:
since we are using 2.7, this seems to be an integer division problem, do you want to make sure you got the right float result? same above and below.

Reply:
I'm using Python 3.5 here. Will mention it at the beginning next time.

Reply:
I am a little bit concerned that our code uses both 2.7 and 3.5. Do you think it's a good idea to stick on 2.7? Seems a trivial work. I built all my codes in a 2.7 venv.

Reply:
Keras is based on Tensorflow, and the latest Tensorflow build no longer support 2.7. That's why I have to use Python 3. Unfortunately it's a standard for neural networks nowadays.